src/Microsoft.DotNet.Archive/IndexedArchive.cs

```diff
@@ -0,0 +1,583 @@
+                    // execute all operations
+                    //foreach(var extractOperation in extractOperations)
+                    extractOperations.AsParallel().ForAll(extractOperation =>
+                    {
+                        extractOperation.DoOperation();
+                        progress.Report("Expanding", Interlocked.Increment(ref opsExecuted), extractOperations.Count);
+                    });
+                }
+            }
+        }
+
+        public void AddExternalDirectory(string externalDirectory)
+        {
+            CheckDisposed();
+            foreach (var externalFile in Directory.EnumerateFiles(externalDirectory, "*", SearchOption.AllDirectories))
+            {
+                AddExternalFile(externalFile);
+            }
+        }
+
+        public void AddExternalFile(string externalFile)
+        {
+            CheckDisposed();
+            using (var fs = File.OpenRead(externalFile))
+            {
+                string hash = GetHash(fs); 
+                // $ prefix indicates that the file is not in the archive and path is relative to an external directory
+                archiveFiles[hash] = new ArchiveFileInfo(null, "$" + hash , hash);
+                externalFiles[hash] = externalFile;
+            }
+        }
+        public void AddDirectory(string sourceDirectory, IProgress<ProgressReport> progress, string destinationDirectory = null)
+        {
+            var sourceFiles = Directory.EnumerateFiles(sourceDirectory, "*", SearchOption.AllDirectories).ToArray();
+            int filesAdded = 0;
+            sourceFiles.AsParallel().ForAll(sourceFile =>
+                {
+                    string destinationPath = sourceFile.Substring(sourceDirectory.Length + 1);
+
+                    if (destinationDirectory != null)
+                    {
+                        destinationPath = Path.Combine(destinationDirectory, destinationPath);
+                    }
+
+                    string extension = Path.GetExtension(sourceFile);
+
+                    if (ZipExtensions.Any(ze => ze.Equals(extension, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        AddZip(sourceFile, destinationPath);
+                    }
+                    else
+                    {
+                        AddFile(sourceFile, destinationPath);
+                    }
+
+                    progress.Report($"Adding {sourceDirectory}", ++filesAdded, sourceFiles.Length);
+                });
+        }
+
+        public void AddZip(string sourceZipFile, string destinationZipFile)
+        {
+            using (var sourceArchive = new ZipArchive(File.OpenRead(sourceZipFile), ZipArchiveMode.Read))
+            {
+                foreach(var entry in sourceArchive.Entries)
+                {
+                    // we can dispose this stream, if AddStream uses it, it will make a copy.
+                    using (var stream = entry.Open())
```

Comment:
The stream returned from entry.Open is backed by a filestream that has a single handle.  We can't have more than one of these open at a time since they would fight over the file handle position.  I would have liked to incrementally build the archive here, but I can't because I don't know the path in the archive until I read the whole file, and I don't know the ordering of the archive until I read all files that are going into it.  As such AddFile holds onto a stream until we are ready to write the archive.  It cannot hold onto the entry stream for the reason I mentioned before.